contents/join.md

```diff
@@ -0,0 +1,140 @@
+But, that doesn't mean that there aren't good use-cases for this join or that you shouldn't use it.
+We can get even more missing data if we use the `crossjoin`.
+This gives the cartesian product of the rows, which is basically multiplication of rows, that is, for every row create a combination with any other row:
+
+```jl
+s = "crossjoin(grades_2020(), grades_2021(); on=:id)"
+sce(s; post=trim_last_n_lines(2))
+```
+
+Oops.
+Since the crossjoin doesn't take the elements in the row into account, we don't need to specify `on` what we want to join:
+
+```jl
+s = "crossjoin(grades_2020(), grades_2021())"
+sce(s; post=trim_last_n_lines(6))
+```
+
+Oops again.
+This is a very common error with DataFrames and joins.
+The tables for the 2020 and 2021 grades have a duplicate column name, namely `:name`.
+We can just pass `makeunique=true` to solve this:
+
+```jl
+s = "crossjoin(grades_2020(), grades_2021(); makeunique=true)"
+sco(s; process=without_caption_label)
+```
+
+So, now, we have one row for each grade from everyone in grades 2020 and grades 2021 datasets.
+For direct queries, such as "who has the highest grade?", the cartesian product is usually not so useful, but for (statistal) queries, it can be.
+
+More useful for scientific projects are the `leftjoin` and `rightjoin`.
+The left join gives all the elements in the _left_ DataFrame:
+
+```jl
+s = "leftjoin(grades_2020(), grades_2021(); on=:name)"
+sco(s; process=without_caption_label)
+```
+
+Here, grades for "Bob" and "Alice" were missing in the grades 2021 table, so that's why there are elements missing.
+The rightjoin does sort of the opposite:
+
+```jl
+s = "rightjoin(grades_2020(), grades_2021(); on=:name)"
+sco(s; process=without_caption_label)
+```
+
+Now, grades in 2020 are missing.
+Note that `leftjoin(A, B) != rightjoin(B, A)`, because the order of the columns will differ.
+For example, compare the output below to the previous output:
+
+```jl
+s = "leftjoin(grades_2021(), grades_2020(); on=:name)"
+sco(s; process=without_caption_label)
+```
+
+Lastly, we have the `semijoin` and `antijoin`.
+The semi join is even more restrictive than the inner join.
+It returns only the elements from the left DataFrame which are in both DataFrames:
+
+```jl
+s = "semijoin(grades_2020(), grades_2021(); on=:name)"
+sco(s; process=without_caption_label)
+```
+
+The opposite of the semi join is the anti join.
+It returns only the elements from the left DataFrame which are **not** in the right DataFrame:
+
+```jl
+s = "antijoin(grades_2020(), grades_2021(); on=:name)"
+sco(s; process=without_caption_label)
+```
+
+## Variable Transformations
+
+```{=comment}
+manipulate variables
+DataFrames.transform
+Ifelse and case_when
+```
+
+In @sec:filter, we saw that `filter` works by taking one or more source columns and filtering it by applying a function.
+In other words, we saw that `source => f::Function` like in, for example, `filter(:name => name -> name == "Alice", df)`.
```

Comment:
DataFrames and bkamins call this "DataFrames minilanguage". Maybe we should introduce this term to the reader?

See here: https://bkamins.github.io/julialang/2020/12/24/minilanguage.html

Reply:
Yeah, I did that later in the text. Here, the full `source => transformation => target` wasn't used. I'll leave this one for you to figure out on your pass over this chapter :yum: